### PR TITLE
feat: query rendering pipeline redesign with SQL dialect integration

### PR DESCRIPTION
**Pull request by jordanfanapour**
## Background

The MJ query rendering stack processes SQL through three stages: **composition** (resolving `{{query:"..."}}` tokens into CTEs), **template evaluation** (Nunjucks `{{ param | filter }}` and `{% if/for %}` blocks), and **pagination** (OFFSET/FETCH or LIMIT/OFFSET). These stages were previously orchestrated inline across two methods in `GenericDatabaseProvider.ts` with duplicated logic between the composition engine and paging engine.

This PR redesigns the pipeline into a single entry point with enforced stage ordering, introduces SQL dialect pluggability throughout the parser, and fixes a production bug in ORDER BY detection that was causing query composition failures.

---

## Bug Fix: ORDER BY Detection Across Template Boundaries

**The production bug:** When a dependency query contained a Nunjucks template expression inside a SQL string literal — like `AND mt.Name = '{{ MembershipType }}'` — MJLexer split this into three tokens (`SQL_TEXT('...= '`)`, `MJ_TEMPLATE_EXPR`, `SQL_TEXT(''\n...ORDER BY...)`). The ORDER BY scanner lost string-literal state at the MJ token boundary, couldn't detect the trailing ORDER BY, and the ORDER BY passed through into the CTE body. SQL Server rejected it with "The ORDER BY clause is invalid in views, inline functions..."

**The fix:** The scanner now carries `inString`, `inBlockComment`, and `inLineComment` state across `SQL_TEXT` token boundaries, matching the existing pattern for `parenDepth`. The scanner also now skips bracket-quoted (`[identifier]`) and double-quoted (`"identifier"`) identifiers to prevent false positives on column aliases like `[Order By Description]`.

---

## Architecture: RenderPipeline

New `RenderPipeline.Run(sql, ctx)` wraps the full rendering sequence with enforced ordering:

```
raw SQL
  │
  ▼ Step 1: Composition
  │  Resolves {{query:"..."}} tokens → CTEs
  │  Strips SQL comments from dependency SQL on load
  │
  ▼ Step 1.5: Comment Stripping
  │  Strips SQL comments from the composed output
  │  (comments serve no runtime purpose; prevents {{ }} in
  │   comment documentation from crashing Nunjucks)
  │
  ▼ Step 2: Nunjucks Template Evaluation
  │  {{ param | filter }} and {% if/for %} blocks
  │
  ▼ Step 3: MaxRows Safety Limit (if specified)
  │
  ▼ Step 4: Pagination (if requested)
  │  OFFSET/FETCH (SQL Server) or LIMIT/OFFSET (PostgreSQL)
  │
  ▼ Final SQL
```

**Diagnostic tracing:** `RenderPipeline.Run()` returns a `RenderTrace` with `AfterComposition`, `AfterTemplates`, and `AfterPaging` SQL snapshots. When SQL Server rejects the output, callers can see what each stage produced to identify which stage mangled the SQL. This replaces the previous failure mode where error diagnosis required reading raw SQL Server error messages.

**Production callers wired:** Both `processQueryParameters` (RunQuery path for saved queries) and `resolveSpecParameters` (TestQuerySQL path for transient specs) now delegate to `RenderPipeline.Run()` instead of calling the three engines inline. No behavioral change — the same composition → template → paging sequence executes, but through a single code path with tracing.

---

## Architecture: Shared ORDER BY Analyzer

The composition engine and paging engine previously had independent implementations of ORDER BY detection and stripping (~300 lines of duplicated logic). The composition version handled SQL+Nunjucks via MJLexer but had the string-literal state bug. The paging version handled clean SQL correctly but couldn't handle MJ tokens.

New `AnalyzeTopLevelOrderBy(sql, dialect)` in `@memberjunction/sql-parser` is the single implementation:
- **Tier 1:** AST parsing via `node-sql-parser` (with Nunjucks placeholder preprocessing if MJ tokens detected)
- **Tier 2:** MJLexer-based scanner with carried lexical state across token boundaries

Both engines delegate to it. The composition engine also runs a post-assembly validation: if any CTE body still has an illegal ORDER BY after stripping, it throws a diagnostic error with the dependency name, category path, and position — instead of letting SQL Server produce the generic error.

---

### Problem
The codebase had ~11 inline `platform === 'postgresql'` switch statements for identifier quoting, paging syntax, parser dialect strings, etc. The `SQLParser` methods accepted raw `parserDialect: string` parameters with `'TransactSQL'` defaults. Adding a new SQL dialect (MySQL, Oracle, SQLite) would require finding and updating every switch statement.

### Solution: `SQLParserDialect` Interface

New interface in `@memberjunction/sql-dialect`:
```typescript
interface SQLParserDialect {
    ParserDialect: string;           // e.g., 'TransactSQL', 'PostgresQL', 'MySQL'
    QuoteIdentifier(name: string): string;  // [name] vs "name" vs `name`
    AllowsOrderByInCTE: boolean;     // false for SQL Server, true for PostgreSQL
    DefaultPagingOrderBy: string;    // '(SELECT NULL)' vs '1'
}
```

`SQLDialect` (the existing 80+ method abstract class) implements this interface. All `SQLParser` public methods (`ParseSQL`, `SqlifyAST`, `ExtractCTEs`, `Astify`, `ExprToSQL`, `ExtractSelectColumns`) now accept `SQLParserDialect` instead of raw strings. All ~60 call sites updated.

### `GetDialect` Factory

Single canonical `GetDialect(platform)` factory in `@memberjunction/sql-dialect` replaces 4 duplicated switch/map implementations across GenericDatabaseProvider (3 copies) and SchemaEngine (1 copy).

### Inline Platform Checks Eliminated

All 11 inline `platform === 'postgresql'` checks in GenericDatabaseProvider migrated to `SQLDialect` method calls:
- Identifier quoting → `dialect.QuoteIdentifier()`
- Parser dialect string → `dialect.ParserDialect`
- Paging clause → `dialect.LimitClause()`
- Default ORDER BY → `dialect.DefaultPagingOrderBy`

---

### Composition IR Types (`@memberjunction/sql-parser`)
Typed intermediate representation for the composition engine: `QueryIR`, `CTENode`, `Fragment` union types (`SQLFragment`, `TemplateExprFragment`, `BlockFragment`, `CommentFragment`, `CompositionRefFragment`). These describe the structure of SQL+MJ token streams for structural manipulation rather than text-level string operations.

### SymbolTable (`@memberjunction/generic-database-provider`)
Guarantees CTE name uniqueness at registration time. When two dependencies each declare an inner CTE with the same name (e.g., both declare `Bridge`), the SymbolTable automatically suffixes the second one (`Bridge__2`) and rewrites references. Replaces the previous `allocatedNames` Set approach.

### Structural Parser (`@memberjunction/sql-parser`)
`ParseToIR()` tokenizes SQL into Fragment sequences via MJLexer, extracts CTE structure, and detects trailing ORDER BY with body splitting. `RenderIR()` serializes back to SQL with `{{ }}`/`{% %}` tokens intact for Nunjucks.

---

## Comment Handling Redesign

**Before:** SQL comments containing `{{ }}` patterns (e.g., documentation like `-- Compose via {{query:"..."}}`) were "escaped" by mangling `{{` → `{ {` before Nunjucks processed them. This produced corrupted comments in the executed SQL.

**After:** SQL comments are **stripped** at two points:
1. **Dependency SQL:** stripped immediately when loaded during composition (protects `substituteStaticParams` from `{{ }}` in dep comments, and prevents "ORDER BY" text in comments from triggering false-positive stripping)
2. **Outer SQL:** stripped by the pipeline between composition and Nunjucks (protects Nunjucks from comment-embedded tokens)

Comments are preserved in `RenderTrace.AfterComposition` for debugging but do not appear in the executed SQL. The `escapeTemplateTokensInComments` method has been removed.

---

## Packages Modified

| Package | Bump | Key changes |
|---------|------|-------------|
| `@memberjunction/core` | minor | New migration file |
| `@memberjunction/sql-dialect` | patch | `SQLParserDialect` interface, `DefaultPagingOrderBy` property, `GetDialect` factory |
| `@memberjunction/sql-parser` | patch | All methods accept `SQLParserDialect`, ORDER BY analyzer + structural parser + composition IR moved here |
| `@memberjunction/generic-database-provider` | patch | `RenderPipeline`, `SymbolTable`, production callers wired, inline platform checks eliminated, comment stripping |
| `@memberjunction/core-entities-server` | patch | Updated to pass dialect objects to `ExtractSelectColumns`; fix pipeline not cleaning up stale Query Fields/Entities |
| `@memberjunction/schema-engine` | patch | `GetDialect` delegates to canonical factory |
| `@memberjunction/react-test-harness` | patch | Updated `ParseSQL` call to pass dialect |
| `@memberjunction/ng-dashboards` | patch | Query browser: tree category dropdown, active folder defaults, per-folder create, Reusable toggle |
| `@memberjunction/ng-core-entity-forms` | patch | Query form: saving indicator, tree category dropdown, Reusable/Cache/Audit fields, fix sub-entity delete |
| `@memberjunction/ng-trees` | patch | Fix `fetchDisplayTextForValue` for branch-only dropdowns |

---

## Test Coverage

**Unit tests:** 67 new tests in `@memberjunction/generic-database-provider`, 2 new in `@memberjunction/sql-parser`

| Category | Tests | What they cover |
|----------|-------|----------------|
| ORDER BY stripping edge cases | 11 | Comment-embedded ORDER BY (Bug A), template in string literal (Bug D), window functions, `{% if %}` conditionals, STRING_AGG WITHIN GROUP, TOP/OFFSET/FOR XML preservation |
| Inner CTE composition | 6 | Basic hoisting, multi-CTE deps, TRY_CAST fallback, Bug A + Bug D patterns |
| Multi-dep composition | 3 | Two deps, CTE name collision (Bug C), partial overlap |
| Full pipeline integration | 2 | All bug patterns combined + paging, conditional blocks + sqlIn |
| Recursive CTEs | 9 | Standalone T-SQL/pgSQL, dep composition, outer recursive CTE merge, recursive + templates/blocks/comments, Bug C + Bug D with recursive |
| Comment stripping | 5 | Dep comments with `{{ }}`, nested deps (2 layers), "ORDER BY" in comments, outer query comments |
| PostgreSQL dialect | 3 | ORDER BY preserved in CTEs, double-quoted identifiers, LIMIT/OFFSET syntax |
| Structural parser + symbol table | 24 | ParseToIR, RenderIR, SymbolTable collision/quoting, IR manipulation |

**Manual testing:** 10 real-world queries created in MJ Explorer across `Pipeline Tests/Standalone`, `Pipeline Tests/Composition`, and `Pipeline Tests/Multi-Dep` categories. Tests cover: simple templates, `{% if/else %}`, `sqlIn` arrays, `{% elif %}` chains, composition with static + pass-through params, inner CTE hoisting + ORDER BY stripping, Bug D string literal pattern, Bug A comment pattern, multi-dep composition, and SQL-level paging.

**CI:** 72/72 turbo build+test tasks pass across all affected packages.

---

## Breaking Change: `SQLParser` Method Signatures

All `SQLParser` public methods that previously accepted `dialect: string = 'TransactSQL'` now require `dialect: SQLParserDialect`. This is an internal monorepo change — all ~60 call sites are updated in this PR. No external consumers are affected.

---

## Query Dashboard & Form UI Improvements

Alongside the rendering pipeline work, several UX issues in the Queries dashboard and entity form were addressed:

- **Tree category dropdown:** Replaced flat `<select>` / `<mj-dropdown>` with `<mj-tree-dropdown>` in both the query browser drawer and the full entity form, showing hierarchical category structure with search
- **Active folder context:** New queries default to the most recently expanded/selected category; per-folder `+` buttons allow creating queries directly in a category
- **Exposed query config fields:** Added Reusable, Cache Enabled, Cache TTL, and Audit Runs toggles to the entity form's Details panel with descriptive hints
- **Saving indicator:** Added spinner overlay and disabled state during save in the query entity form, with `[IsSaving]` passed to the toolbar
- **Sub-entity delete fix:** Parameter/Field/Entity delete now reloads a fresh entity copy via `Metadata.GetEntityObject()` before deleting, with `LatestResult.CompleteMessage` surfaced in error notifications
- **Tree dropdown fix:** `fetchDisplayTextForValue` now falls back to `BranchConfig.EntityName` when no `LeafConfig` is present (branch-only dropdowns)
- **Pipeline cleanup fix:** Added `RemoveAllRecords` calls for Query Fields and Query Entities when extraction produces no results — previously stale records from a prior SQL version were orphaned (9 regression tests added)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/GenericDatabaseProvider/src/renderPipeline.ts [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+ * 3. **Paging** detects ORDER BY and injects OFFSET/FETCH or LIMIT/OFFSET.
+ *    Must run AFTER Nunjucks because it needs to detect whether
+ *    `{% if SortField %} ORDER BY ... {% endif %}` resolved to an actual
+ *    ORDER BY or not — that depends on runtime parameter values.
+ *
+ * The `Trace` field on the result captures the SQL state after each pass,
+ * enabling diagnosis when SQL Server rejects the output.
+ */
+export class RenderPipeline {
+    private static readonly compositionEngine = new QueryCompositionEngine();
```
Comment (AN-BC) [inline packages/GenericDatabaseProvider/src/renderPipeline.ts]: perhaps should use a true global singleton via `BaseSingleton` since static isn't guaranteed to be truly global across the process space

File: packages/React/test-harness/src/lib/constraint-validators/sql-where-clause-validator.ts [hunk 2]
```diff
@@ -45,6 +45,9 @@ import { BaseConstraintValidator } from './base-constraint-validator';
 import { ValidationContext } from './validation-context';
 import { PropValueExtractor } from '../prop-value-extractor';
 import { SQLParser } from '@memberjunction/sql-parser';
+import { SQLServerDialect } from '@memberjunction/sql-dialect';
+
+const tsqlDialect = new SQLServerDialect();
```
Comment (AN-BC) [inline packages/React/test-harness/src/lib/constraint-validators/sql-where-clause-validator.ts]: we are assuming SQL Server here, shouldn't we have this configurable so the caller of the test harness can control this?

**Review by AN-BC — COMMENTED** [review]
LGTM overall, some minor stuff noted below to fix, after that I think you can merge assuming testing has gone well.

**Comment by jordanfanapour** [comment]
**Re: compositionEngine singleton (line 113)** — Fixed. Removed the static instance. `QueryCompositionEngine` is stateless (no instance fields), so `RenderPipeline` now creates a new instance per call. No shared mutable state across the process space.

**Re: hardcoded SQLServerDialect in test harness (line 50)** — Fixed. Added `SqlWhereClauseValidator.SetDialect(platform)` static method so the test harness caller can configure the dialect before running validations. Defaults to SQL Server for backward compatibility.
